### PR TITLE
feat(user): edit card details

### PR DESCRIPTION
**Pull request by kmitrovv**
## Description

https://www.notion.so/Viewer-Edit-card-details-on-web-app-d74a003781b5446dbb9dee892d5ffca9

### Steps completed:

According to our definition of done, I have completed the following steps:

- [ ] Acceptance criteria met
- [ ] Unit tests added
- [ ] Docs updated (including config and env variables)
- [ ] Translations added
- [ ] UX tested
- [ ] Browsers / platforms tested
- [ ] Rebased & ready to merge without conflicts
- [ ] Reviewed own code

### CODE REVIEW

File: src/components/EditCardPaymentForm/EditCardPaymentForm.tsx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+const EditCardPaymentForm: React.FC<Props> = ({ onCancel, setUpdatingCardDetails }) => {
+  const { t } = useTranslation('account');
+  const updateCard = useMutation(updateCardDetails);
+  const { activePayment } = useAccountStore(({ activePayment }) => ({ activePayment }), shallow);
+  const paymentData = useForm(
+    { cardholderName: '', cardNumber: '', cardExpiry: '', cardCVC: '', cardExpMonth: '', cardExpYear: '' },
+    async () => {
+      setUpdatingCardDetails(true);
+      await updateCard.mutateAsync({
```
Comment (naumovski-filip) [inline src/components/EditCardPaymentForm/EditCardPaymentForm.tsx]: Can we use onSettled here instead and call onCancel inside it?

File: src/components/EditCardPaymentForm/EditCardPaymentForm.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        cardName: paymentData.values.cardholderName,
+        cardNumber: paymentData.values.cardNumber.replace(/\s+/g, ''),
+        cvc: parseInt(paymentData.values.cardCVC),
+        expMonth: parseInt(paymentData.values.cardExpMonth),
+        expYear: parseInt(paymentData.values.cardExpYear),
+        currency: activePayment?.currency || '',
+      });
+      onCancel();
```
Comment (filip-naumovski) [inline src/components/EditCardPaymentForm/EditCardPaymentForm.tsx]:
```suggestion
      updateCard.mutate(
        {
          cardName: paymentData.values.cardholderName,
          cardNumber: paymentData.values.cardNumber.replace(/\s+/g, ''),
          cvc: parseInt(paymentData.values.cardCVC),
          expMonth: parseInt(paymentData.values.cardExpMonth),
          expYear: parseInt(paymentData.values.cardExpYear),
          currency: activePayment?.currency || '',
        },
        {
          onSettled: () => onCancel(),
        },
      );
```

File: src/components/EditCardPaymentForm/EditCardPaymentForm.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+          <CreditCardCVCField
+            value={paymentData?.values?.cardCVC}
+            onBlur={paymentData?.handleBlur}
+            onChange={paymentData?.handleChange}
+            error={paymentData?.errors?.cardCVC}
+          />
+        </div>
+      </div>
+      <div>
+        <Button label={t('checkout.save')} variant="contained" onClick={paymentData.handleSubmit as () => void} color="primary" size="large" fullWidth />
```
Comment (naumovski-filip) [inline src/components/EditCardPaymentForm/EditCardPaymentForm.tsx]:
```suggestion
        <Button disabled={updateCard.isLoading} label={t('checkout.save')} variant="contained" onClick={paymentData.handleSubmit as () => void} color="primary" size="large" fullWidth />
```

File: src/components/Payment/Payment.tsx [hunk 4]
```diff
@@ -168,6 +175,7 @@ const Payment = ({
                 <TextField label={t('user:payment.expiry_date')} value={activePaymentDetail.paymentMethodSpecificParams.cardExpirationDate} editing={false} />
                 <TextField label={t('user:payment.security_code')} value={'******'} editing={false} />
               </div>
+              <Button label={t('Edit card')} variant="outlined" onClick={onEditCardDetailsClick} />
```
Comment (ChristiaanScheermeijer) [inline src/components/Payment/Payment.tsx]:
```suggestion
              <Button label={t('account:payment.edit_card')} variant="outlined" onClick={onEditCardDetailsClick} />
```

File: src/stores/AccountController.ts [hunk 5]
```diff
@@ -342,6 +342,35 @@ export const updateSubscription = async (status: 'active' | 'cancelled'): Promis
...
+}) => {
+  return await useAccount(async ({ customerId }) => {
+    return await useService(async ({ subscriptionService, sandbox = true }) => {
+      const response = await subscriptionService?.updateCardDetails({ cardName, cardNumber, cvc, expMonth, expYear, currency }, sandbox);
+      const activePayment = (await subscriptionService?.getActivePayment({ sandbox, customerId })) || null;
+      useAccountStore.setState({ loading: true });
+      useAccountStore.setState({
+        loading: false,
+        activePayment,
+      });
```
Comment (ChristiaanScheermeijer) [inline src/stores/AccountController.ts]: Is this correct (2x setState)?

File: src/components/EditForm/EditCardDetailsForm.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  submitting: boolean;
+  onCancel: () => void;
+};
+
+const EditCardDetailsForm: React.FC<Props> = ({ renderPaymentMethod, submitting, onCancel }) => {
+  const { t } = useTranslation('common');
+  return (
+    <div>
+      <div className={styles.title}>
+        <h1>Add card details</h1>
```
Comment (ChristiaanScheermeijer) [inline src/components/EditForm/EditCardDetailsForm.tsx]:
```suggestion
        <h1>t('user:payments.add_card_details')</h1>
```

File: src/components/EditForm/EditCardDetailsForm.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import LoadingOverlay from '#components/LoadingOverlay/LoadingOverlay';
+
+type Props = {
+  renderPaymentMethod?: () => JSX.Element | null;
+  submitting: boolean;
+  onCancel: () => void;
+};
+
+const EditCardDetailsForm: React.FC<Props> = ({ renderPaymentMethod, submitting, onCancel }) => {
+  const { t } = useTranslation('common');
```
Comment (ChristiaanScheermeijer) [inline src/components/EditForm/EditCardDetailsForm.tsx]:
```suggestion
  const { t } = useTranslation(['common', 'user']);
```

File: src/components/EditCardPaymentForm/EditCardPaymentForm.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+      }
+    }
+    //eslint-disable-next-line
+  }, [paymentData.values.cardExpiry]);
+
+  return (
+    <div className={styles.paymentForm}>
+      <div>
+        <TextField
+          label="Cardholder name"
```
Comment (ChristiaanScheermeijer) [inline src/components/EditCardPaymentForm/EditCardPaymentForm.tsx]: We should translate this field as well

File: src/components/EditForm/EditCardDetailsForm.tsx [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  return (
+    <div>
+      <div className={styles.title}>
+        <h1>Add card details</h1>
+      </div>
+      <div>{renderPaymentMethod ? renderPaymentMethod() : null}</div>
+      <Button onClick={onCancel} className={styles.cancelButton} fullWidth label={t('confirmation_dialog.close')} />
```
Comment (ChristiaanScheermeijer) [inline src/components/EditForm/EditCardDetailsForm.tsx]:
```suggestion
      <Button onClick={onCancel} className={styles.cancelButton} fullWidth label={t('common:confirmation_dialog.close')} />
```

File: src/services/inplayer.subscription.service.ts [hunk 10]
```diff
@@ -90,8 +94,17 @@ export const updateSubscription: UpdateSubscription = async ({ offerId, unsubscr
...
 
-const formatCardDetails = (card: Card): PaymentDetail => {
-  const { number, exp_month, exp_year, card_name, card_type, account_id } = card;
+export const updateCardDetails: UpdateCardDetails = async ({ cardName, cardNumber, cvc, expMonth, expYear, currency }) => {
+  try {
+    const response = await InPlayer.Payment.setDefaultCreditCard({ cardName, cardNumber, cvc, expMonth, expYear, currency });
+
+    return { responseData: response.data, errors: [] };
+  } catch {
+    throw new Error('Failed to update card details');
```
Comment (ChristiaanScheermeijer) [inline src/services/inplayer.subscription.service.ts:103]: Is this error shown in the form?

Comment (borkopetrovicc) [reply src/services/inplayer.subscription.service.ts:103]: No it is not

File: src/stores/AccountController.ts (hunk 5 above)
Comment (borkopetrovicc) [reply src/stores/AccountController.ts]: 935748a0862d1a60e85231924c272b4981f42060

File: src/components/EditCardPaymentForm/EditCardPaymentForm.tsx (hunk 8 above)
Comment (borkopetrovicc) [reply src/components/EditCardPaymentForm/EditCardPaymentForm.tsx]: 935748a0862d1a60e85231924c272b4981f42060

**Review by ChristiaanScheermeijer — APPROVED** [review]
Lgtm!

**Pull request merged**